### PR TITLE
ADC: STM32: Add device tree config for clock source with prescaler

### LINKED ISSUE #40955: ADC: STM32: Add device tree config for clock source
Looking here https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_stm32.c#L897, you can see that ADC clock source is hard-coded to `LL_ADC_CLOCK_ASYNC_DIV4`. I'd like to see some config added to the device tree to allow one to specify this.

I really need some help on what would be a good device tree addition to represent this.

### PR DESCRIPTION
**Pull request by FRASTM**
This PR adds a clock prescaler to the ADC of the stm32 mcus

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/40955

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -236,6 +236,45 @@ static const uint32_t table_samp_time[] = {
...
+	LL_ADC_CLOCK_ASYNC_DIV64,
+	LL_ADC_CLOCK_ASYNC_DIV128,
+	LL_ADC_CLOCK_ASYNC_DIV256,
+#endif
+};
+
+/* Bugfix for STM32G4 HAL */
+#if !defined(ADC_CHANNEL_TEMPSENSOR)
+#define ADC_CHANNEL_TEMPSENSOR ADC_CHANNEL_TEMPSENSOR_ADC1
+#endif
```
Comment (pdietl) [inline drivers/adc/adc_stm32.c:276]: Why is this added back in?

File: dts/bindings/adc/st,stm32-adc.yaml [hunk 2]
```diff
@@ -27,6 +27,14 @@ properties:
...
       required: true
 
+    st,prescaler:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default is 0 (not divided)
+
```
Comment (pdietl) [inline dts/bindings/adc/st,stm32-adc.yaml:37]: Does it make sense to use an enum type here?

File: dts/bindings/adc/st,stm32-adc.yaml [hunk 3]
```diff
@@ -27,6 +27,14 @@ properties:
     pinctrl-names:
       required: true
 
+    st,prescaler:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default is 0 (not divided)
```
Comment (mbolivar-nordic) [inline dts/bindings/adc/st,stm32-adc.yaml:36]: This default needs a justification as per https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties

**Review by mbolivar-nordic — DISMISSED** [review]

File: dts/bindings/adc/st,stm32-adc.yaml (hunk 2 above)
Comment (FRASTM) [reply dts/bindings/adc/st,stm32-adc.yaml:37]: adding `Valid range: 0 - 15`

File: dts/bindings/adc/st,stm32-adc.yaml (hunk 3 above)
Comment (FRASTM) [reply dts/bindings/adc/st,stm32-adc.yaml:36]: Ok, default is the reset value

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: removing this code

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
 };
 #endif
 
+/* table to define the LL ADC clock prescaler from the DTS */
```
Comment (pdietl) [inline drivers/adc/adc_stm32.c:239]: We should add a static assert that `config->clock_prescaler < ARR_SIZE(adc_clock_prescaler)`

**Review by pdietl — CHANGES_REQUESTED** [review]

File: dts/bindings/adc/st,stm32-adc.yaml (hunk 2 above)
Comment (FRASTM) [reply dts/bindings/adc/st,stm32-adc.yaml]: adding `Valid range: 0 - 15`

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (FRASTM) [reply drivers/adc/adc_stm32.c:239]: with ASSERT_NO_MSG

File: drivers/adc/adc_stm32.c [hunk 5]
```diff
@@ -921,6 +957,9 @@ static int adc_stm32_init(const struct device *dev)
 		return -EIO;
 	}
 
+	__ASSERT_NO_MSG(config->clock_prescaler < ARRAY_SIZE(adc_clock_prescaler));
```
Comment (pdietl) [inline drivers/adc/adc_stm32.c:960]: This is still a runtime assert. This could be found at compile time.

**Review by pdietl — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 5 above)
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: changing to BUILD_ASSERT

File: dts/bindings/adc/st,stm32-adc.yaml [hunk 6]
```diff
@@ -27,6 +27,31 @@ properties:
...
+        - 9
+        - 10
+        - 11
+        - 12
+        - 13
+        - 14
+        - 15
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default reset value is 0 (not divided)
```
Comment (erwango) [inline dts/bindings/adc/st,stm32-adc.yaml:53]: Remove "Could be in range [0 .. 0xF] ; "

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
Please update commit author: `author email (Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>) needs to match one of the signed-off-by entries.`

`git amend --author="François ...."`

**Comment by FRASTM** [comment]
This part of the stm32 adc driver in the `adc_stm32_init` function requires more changes because the ADC clock can now hav different prescaler value 
including Synchronous or asynchronous division factor. That should be handled a particular call to LL_ADC_SetCommonClock and/or LL_ADC_SetClock depending on the stm32 soc

File: dts/bindings/adc/st,stm32-adc.yaml (hunk 6 above)
Comment (FRASTM) [reply dts/bindings/adc/st,stm32-adc.yaml:53]: removed

**Comment by FRASTM** [comment]
rebase on

File: drivers/adc/adc_stm32.c [hunk 7]
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
...
+	LL_ADC_CLOCK_ASYNC_DIV6,
+	LL_ADC_CLOCK_ASYNC_DIV8,
+	LL_ADC_CLOCK_ASYNC_DIV10,
+	LL_ADC_CLOCK_ASYNC_DIV12,
+	LL_ADC_CLOCK_ASYNC_DIV16,
+	LL_ADC_CLOCK_ASYNC_DIV32,
+	LL_ADC_CLOCK_ASYNC_DIV64,
+	LL_ADC_CLOCK_ASYNC_DIV128,
+	LL_ADC_CLOCK_ASYNC_DIV256,
+#endif
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:271]: These are also supported by most other series: CONFIG_SOC_SERIES_STM32L4X || CONFIG_SOC_SERIES_STM32L5X || CONFIG_SOC_SERIES_STM32WBX || CONFIG_SOC_SERIES_STM32G0X || CONFIG_SOC_SERIES_STM32G4X || CONFIG_SOC_SERIES_STM32H7X

**Review by erwango — CHANGES_REQUESTED** [review]
We need to revisit this change as it mixes CKMODE and PRESC and binding doesn't reflect sufficiently allow to describe socs capabilities/limitations

File: drivers/adc/adc_stm32.c [hunk 8]
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
...
+};
+
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:273]:
```suggestion
/* table to define the LL ADC clock prescaler from the DTS */

#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
	defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32WLX)
        defined(STM32F3X_ADC_V1_1) || \
	defined(CONFIG_SOC_SERIES_STM32L4X) || \
	defined(CONFIG_SOC_SERIES_STM32L5X) || \
	defined(CONFIG_SOC_SERIES_STM32WBX) || \
	defined(CONFIG_SOC_SERIES_STM32G0X) || \
	defined(CONFIG_SOC_SERIES_STM32G4X) || \
	defined(CONFIG_SOC_SERIES_STM32H7X)
#define z_clk(v)	LL_ADC_CLOCK_SYNC_PCLK_##v
#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
	defined(CONFIG_SOC_SERIES_STM32U5X)
#define z_clk(v)	LL_ADC_CLOCK_ASYNC_##v
#endif

#define adc_clock_prescaler(v) z_clk(v)
```

Then you can do
```
LL_ADC_SetClock(adc, adc_clock_prescaler[config->clock_prescaler]);
```

File: drivers/adc/adc_stm32.c (hunk 7 above)
Comment (FRASTM) [reply drivers/adc/adc_stm32.c:270]: CONFIG_SOC_SERIES_STM32U5X and CONFIG_SOC_SERIES_STM32L0X only have LL_ADC_CLOCK_ASYNC_DIVx  for their common clock config. 
Soc series that have LL_ADC_CLOCK_SYNC_PCLK_DIVx and also LL_ADC_CLOCK_ASYNC_DIVx to set their LL_ADC_SetCommonClock, would require a prescaler for Synchronous clock and one for Asynchronous clock.

File: drivers/adc/adc_stm32.c [hunk 9]
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
 };
 #endif
 
+/* table to define the LL ADC clock prescaler from the DTS */
+static const uint32_t adc_clock_prescaler[] = {
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: removed for stm32F0x

File: drivers/adc/adc_stm32.c [hunk 10]
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
...
+#elif defined(STM32F3X_ADC_V1_1) || \
+	defined(CONFIG_SOC_SERIES_STM32L4X) || \
+	defined(CONFIG_SOC_SERIES_STM32L5X) || \
+	defined(CONFIG_SOC_SERIES_STM32WBX) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G4X) || \
+	defined(CONFIG_SOC_SERIES_STM32H7X)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: merged

File: drivers/adc/adc_stm32.c [hunk 11]
```diff
@@ -973,11 +1011,11 @@ static int adc_stm32_init(const struct device *dev)
 	defined(CONFIG_SOC_SERIES_STM32G4X) || \
 	defined(CONFIG_SOC_SERIES_STM32H7X)
 	LL_ADC_SetCommonClock(__LL_ADC_COMMON_INSTANCE(adc),
-			      LL_ADC_CLOCK_SYNC_PCLK_DIV4);
+			      adc_clock_prescaler[config->clock_prescaler]);
 #elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
 	defined(CONFIG_SOC_SERIES_STM32U5X)
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 8 above)
Comment (str4t0m) [reply drivers/adc/adc_stm32.c]: I think that specifying the actual clock-divider via dts instead of the register value would be more intuitive for the user.
And then you could use the macro recommended by erwango instead of using an array to translate the register value back and forth.

Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: Defining the st,prescaler property  directly with the divider value to be 1, 2, 4, 6, 8, 10, 12, 16, 32, 64,128,256  is possible in the dts. The related adc_clock_prescaler will depend on the soc serie 
either LL_ADC_CLOCK_SYNC_PCLK_##v or LL_ADC_CLOCK_ASYNC_##v 
Except for the **stm32F0** where default 1 is LL_ADC_CLOCK_ASYNC and 2 or 4 are LL_ADC_CLOCK_SYNC_PCLK_DIV_##v

Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: then configuring with  `LL_ADC_SetClock(adc, adc_clock_prescaler(config->clock_prescaler));`

Comment (erwango) [reply drivers/adc/adc_stm32.c]: > However with` #define adc_clock_prescaler(v) z_clock_prescaler(v)` "v" must be like "DT_INST_PROP(0, st_prescaler)" and cannot be "config->clock_prescaler" (-->`error: 'LL_ADC_CLOCK_ASYNC_DIVconfig' undeclared`)

Right, since this rely on pre-processing macros 'v' should be available at build time, which is not the case for config->clock_prescaler.

Anyway, we should find another way to allow the user to provide the prescaler value instead of an array rank.

Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: Using COND_CODE_1, it is possible to build a LL_ADC_CLOCK_ASYNC_DIVn or LL_ADC_CLOCK_SYNC_PCLK_DIVn
value from the st,prescaler property (based on string concatenation)

**Comment by FRASTM** [comment]
- rebase on 5897de1e535816247e309348cbe29760361a28b8
- define the st,prescaler as a string and build the clock_prescaler using concatenation.

Example for stm32u5x, to set LL_ADC_CLOCK_ASYNC_DIV6, the dts must define the property as follows: 
```
&adc1 {
	st,prescaler = "6";
};
```

**Comment by FRASTM** [comment]
The stm32F1x has no prescaler, its config->clock_prescaler is reset

**Comment by FRASTM** [comment]
See https://github.com/zephyrproject-rtos/zephyr/pull/48938
